### PR TITLE
STYLE: Explicitly defaulted default-constructors Vector Accessor classes

### PR DESCRIPTION
**Pull request by N-Dekker**
Defaulted (`= default`) the default-constructors of Vector Accessor classes,
`DefaultVectorPixelAccessor`, `DefaultVectorPixelAccessorFunctor`, and
`VectorImageNeighborhoodAccessorFunctor`.

Initialized the `m_Begin` data member of both `Functor` classes by a C++11
in-class default member initializer, `{nullptr}`.

Note: this commit ensures that these default-constructors will usually get a
non-throwing exception specification (implicitly). For example,
`noexcept(itk::DefaultVectorPixelAccessor<int>{})` was `false` before this
commit, and will be `true` after this commit.

### CODE REVIEW

File: Modules/Core/Common/include/itkDefaultVectorPixelAccessorFunctor.h [hunk 1]
```diff
@@ -106,7 +106,7 @@ class DefaultVectorPixelAccessorFunctor
 
 private:
   PixelAccessorType  m_PixelAccessor;    // The pixel accessor
-  InternalPixelType *m_Begin;            // Begin of the buffer
+  InternalPixelType *m_Begin{nullptr};            // Begin of the buffer
```
Comment (phcerdan) [inline Modules/Core/Common/include/itkDefaultVectorPixelAccessorFunctor.h]: Align comment

File: Modules/Core/Common/include/itkVectorImageNeighborhoodAccessorFunctor.h [hunk 2]
```diff
@@ -129,7 +128,7 @@ class VectorImageNeighborhoodAccessorFunctor
   VectorLengthType m_VectorLength{0};
   VectorLengthType m_OffsetMultiplier{0}; // m_OffsetMultiplier = m_VectorLength-1
                                        // (precomputed for speedup).
-  InternalPixelType *m_Begin;          // Begin of the buffer.
+  InternalPixelType *m_Begin{nullptr};          // Begin of the buffer.
```
Comment (phcerdan) [inline Modules/Core/Common/include/itkVectorImageNeighborhoodAccessorFunctor.h]: Align comments

**Comment by N-Dekker** [comment]
@phcerdan wrote:
> Isn't the constructor already treated as noexcept?
I guess no hurt in adding noexcept, but if this is implicitly done/deduced by the compiler, no need on this. Why adding noexcept here and not in all the constructors in the library?

Good question! However, I'm prepared  :-)  I assume you're specifically referring to the default-constructors, right?

First of all, I think we should _only_ add a `noexcept` specifier (or `noexcept(true)`)  to those functions that would never ever throw. (More precisely: we should ensure that such functions would never have any unhandled exception.) But I guess we all agree on that already.

Now it appears that for a defaulted (`= default`) member function that has a `noexcept` specifier, the compiler checks if it can actually generate a non-throwing implementation! More specifically: the compiler _only_ does such a check when the defaulted member function has a `noexcept` specifier! Please check: https://rextester.com/UNIR24573

    class ThrowingConstructor
    {
    public:
      ThrowingConstructor() { throw 42; }
    };

    class DefaultedConstructor
    {
      ThrowingConstructor data;
    public:
      DefaultedConstructor() = default;
    };

    class NoexceptDefaultedConstructor
    {
      ThrowingConstructor data;
    public:
      // error: exception specification of explicitly defaulted default constructor
      // does not match the calculated one
      NoexceptDefaultedConstructor() noexcept = default;
    };

    int main()
    {
      DefaultedConstructor x;
      NoexceptDefaultedConstructor y;
    }

Do you agree that that's a good reason to add a `noexcept` specifier to these defaulted default-constructors, of this PR?

File: Modules/Core/Common/include/itkDefaultVectorPixelAccessorFunctor.h [hunk 3]
```diff
@@ -77,7 +77,7 @@ class DefaultVectorPixelAccessorFunctor
   }
 
 
-  DefaultVectorPixelAccessorFunctor () : m_Begin(nullptr) {}
+  DefaultVectorPixelAccessorFunctor() = default;
```
Comment (dzenanz) [inline Modules/Core/Common/include/itkDefaultVectorPixelAccessorFunctor.h]: This is missing `ITK_NOEXCEPT`. Is that on purpose?

Comment (N-Dekker) [reply Modules/Core/Common/include/itkDefaultVectorPixelAccessorFunctor.h]: @dzenanz I would have liked to add `ITK_NOEXCEPT` to this specific default-constructor as well, but I hesitated, because it was not 100% clear to me if its data member `DefaultVectorPixelAccessorFunctor::m_PixelAccessor` could possibly have a throwing default-constructor. So I thought: when in doubt, don't noexcept!

**Comment by phcerdan** [comment]
> Do you agree that that's a good reason to add a noexcept specifier to these defaulted default-constructors, of this PR?

We have had this conversation before, noexcept is an optimization flag, not something that has to be added everywhere. Your example is a great exercise showing what happens when a data member throws in its constructor. However, it does not cover my honest question: is there such a throwing data member in this class?
If it is not throwing, the compiler will do the exact same optimization with or without noexcept. noexcept might help to read/think about the code, which is important, but at the same time, it would generate some kind of viral expansion, with the need of putting noexcept everywhere.

In other words, the compiler only does the check when noexcept is specified, however, not putting noexcept doesn't forbid the compiler to do the same optimizations.

Why is this class special? If it is a complex, and putting noexcept helps, I am in.

Other conversation would be to add noexcept as generous as possible. Which might not be a bad thing, but would require consistency among the database.

**Comment by N-Dekker** [comment]
For non-defaulted functions, it's sometimes risky to add a `noexcept` specifier, because a misplaced  `noexcept` at a non-defaulted function might cause program termination. However, for a defaulted (=default) function, the compiler checks that it's OK to add `noexcept`. The advantages are then:
 
- the non-throwing guarantee provided by the function is documented automatically
- the compiler ensures that with future changes of the class, the function cannot _accidentally_ start throwing an exception

These Pixel Accessor classes are used within a performance critical context (low level pixel wise operations), so for these, I think it could be beneficial to add `noexcept`.

These are just my thoughts for now. However, if you think it's better to postpone a decision on the proposed `noexcept` specifiers of this PR, I wouldn't mind removing them from the PR.

**Comment by hjmjohnson** [comment]
@N-Dekker What is the status of this PR?  Does it need to be rebased on top of #382?  Are there outstanding issues?

**Comment by N-Dekker** [comment]
@hjmjohnson I think the PR is fine (except for a small misalignment of some comments), but Pablo asked me to remove `ITK_NOEXCEPT` that I added to the constructors, as discussed here. Now I'm just studying the pro's and cons of `noexcept` specifiers. While I still think the PR is fine, I wouldn't mind if this PR gets postponed for a while. It's no show-stopper.

**Comment by hjmjohnson** [comment]
@N-Dekker Can you edit the PR to make it a WIP until it is ready to go?  Otherwise I keep looking at it to see if it can be pushed forward.

The longer these items persist on the list, the more effort is taken to explicitly ignore them for while.  The process of ignoring is also time consuming.

**Comment by N-Dekker** [comment]
@hjmjohnson @thewtex What is the preferred way on GitHub to mark a PR as "postponed", for me as submitter of the PR? I would rather _not_ amend the commit, just to mark it "postponed".  The label `status:Backlog` seems appropriate, but a normal (non-admin) user cannot add labels to a PR, apparently. Adding "WIP" to the title of the PR (rather than the commit) does not prevent an accidental merge.

**Comment by hjmjohnson** [comment]
Adding WIP does make it much harder to commit.  We have checks in place that require administrative user access (elevated) AND also has an additional level of confirmation.  

WIP has the same effect as failing the Dashboard.

**Comment by hjmjohnson** [comment]
It is easy to add, and easy to remove right from the GitHub GUI.

**Comment by N-Dekker** [comment]
Thanks, @hjmjohnson 
> Adding WIP does make it much harder to commit. We have checks in place that require administrative user access (elevated) AND also has an additional level of confirmation.

> WIP has the same effect as failing the Dashboard.

Are you sure this is about adding WIP to the PR title (rather than the commit title)? I do see some closed pull requests with "WIP" in the title, at:
https://github.com/InsightSoftwareConsortium/ITK/pulls?q=is%3Apr+is%3Aclosed

Anyway, if a "WIP" in the PR title is helpful, it's fine by me.

**Comment by hjmjohnson** [comment]
image

Note the yellow dot.  Without the green checkmark, this is harder to commit.

**Comment by N-Dekker** [comment]
The force-push I just did removes all `ITK_NOEXCEPT` specifiers from this pull request, hoping to make it acceptable for ITK5.  @phcerdan  Pablo, is it OK to you now?

**Comment by hjmjohnson** [comment]
@phcerdan Could you please review your requested changes?  It would be nice to have this PR completed.
